File: lwk_bindings/tests/bindings/basics.py
from lwk import *

node = LwkTestEnv() # launch electrs and elementsd

# ANCHOR: generate-signer
mnemonic = Mnemonic.from_random(12)
network = Network.testnet()
network = Network.regtest_default()  # ANCHOR: ignore
signer = Signer(mnemonic, network)
# ANCHOR_END: generate-signer

# ANCHOR: get-xpub
xpub = signer.keyorigin_xpub(Bip.new_bip84())
# ANCHOR_END: get-xpub

# ANCHOR: wollet
desc = signer.wpkh_slip77_descriptor()
wollet = Wollet(network, desc, datadir=None)
# ANCHOR_END: wollet

# ANCHOR: address
addr = wollet.address(None)

# ANCHOR_END: address
client = ElectrumClient(node.electrum_url(), tls=False, validate_domain=False)

sats = 100000
txid = node.send_to_address(addr.address(), sats, asset=None)
wollet.wait_for_tx(txid, client)

# ANCHOR: txs
txs = wollet.transactions()
balance = wollet.balance()
# ANCHOR_END: txs

# ANCHOR: electrum_client
# Create electrum client with custom URL
client = ElectrumClient("blockstream.info:995", tls=True, validate_domain=True)

# Or use the default electrum client for the network
default_client = Network.mainnet().default_electrum_client()
# ANCHOR_END: electrum_client

# ANCHOR: esplora_client
url = "https://blockstream.info/liquid/api"
client = EsploraClient(url, Network.mainnet())
# ANCHOR_END: esplora_client

# ANCHOR: waterfalls_client
url = "https://waterfalls.liquidwebwallet.org/liquid/api"
client = EsploraClient.new_waterfalls(url, Network.mainnet())
# ANCHOR_END: waterfalls_client

# ANCHOR: client
url = "https://blockstream.info/liquidtestnet/api"
client = EsploraClient(url, network)
client = ElectrumClient(node.electrum_url(), tls=False, validate_domain=False)  # ANCHOR: ignore

update = client.full_scan(wollet)
wollet.apply_update(update)
# ANCHOR_END: client

# Receive some funds
address = node.get_new_address()
sats = 1000
lbtc = network.policy_asset()

# ANCHOR: tx
b = network.tx_builder()
b.add_recipient(address, sats, lbtc)
pset = b.finish(wollet)
# ANCHOR_END: tx

# ANCHOR: pset-details
details = wollet.pset_details(pset)
# ANCHOR_END: pset-details

# ANCHOR: sign
pset = signer.sign(pset)
# ANCHOR_END: sign

# ANCHOR: broadcast
tx = pset.finalize()
txid = client.broadcast(tx)

# (optional)
wollet.apply_transaction(tx)
# ANCHOR_END: broadcast

File: lwk_bindings/go/basics.go
package main

import (
	"fmt"
	"log"
	"lwk-go/lwk"
)

func main() {
	node := lwk.NewLwkTestEnv()

	// ANCHOR: generate-signer
	mnemonic, err := lwk.MnemonicFromRandom(12)
	if err != nil {
		log.Fatal(err)
	}
	network := lwk.NetworkTestnet()
	network = lwk.NetworkRegtestDefault() // ANCHOR: ignore
	signer, err := lwk.NewSigner(mnemonic, network)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: generate-signer

	// ANCHOR: get-xpub
	xpub, err := signer.KeyoriginXpub(lwk.BipNewBip84())
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: get-xpub
	fmt.Println("Xpub:", xpub)

	// ANCHOR: wollet
	desc, err := signer.WpkhSlip77Descriptor()
	if err != nil {
		log.Fatal(err)
	}
	wollet, err := lwk.NewWollet(network, desc, nil)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: wollet

	// ANCHOR: address
	addr, err := wollet.Address(nil)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: address

	elClient, err := lwk.ElectrumClientFromUrl(node.ElectrumUrl())
	if err != nil {
		panic(err)
	}
	sats := uint64(100000)
	txid := node.SendToAddress(addr.Address(), sats, nil)
	walletTx, err := wollet.WaitForTx(txid, elClient)
	if err != nil {
		panic(err)
	}
	policyAsset := network.PolicyAsset()
	if walletTx.Balance()[policyAsset] != int64(sats) {
		panic("unexpected balance")
	}

	// ANCHOR: txs
	txs, err := wollet.Transactions()
	if err != nil {
		log.Fatal(err)
	}
	balance, err := wollet.Balance()
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: txs
	fmt.Printf("Transactions: %d, balance: %d sats\n", len(txs), balance[policyAsset])

	// ANCHOR: electrum_client
	// Create electrum client with custom URL
	electrumClient, err := lwk.NewElectrumClient("blockstream.info:995", true, true)
	if err != nil {
		log.Fatal(err)
	}

	// Or use the default electrum client for the network
	defaultClient, err := lwk.NetworkMainnet().DefaultElectrumClient()
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: electrum_client
	_ = electrumClient
	_ = defaultClient

	// ANCHOR: esplora_client
	esploraClient, err := lwk.NewEsploraClient("https://blockstream.info/liquid/api", lwk.NetworkMainnet())
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: esplora_client
	_ = esploraClient

	// ANCHOR: waterfalls_client
	waterfallsClient, err := lwk.NewWaterfallsClient("https://waterfalls.liquidwebwallet.org/liquid/api", lwk.NetworkMainnet())
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: waterfalls_client
	_ = waterfallsClient

	// ANCHOR: client
	url := "https://blockstream.info/liquidtestnet/api"
	client, err := lwk.NewEsploraClient(url, network)
	if err != nil {
		log.Fatal(err)
	}
	client, err = lwk.NewEsploraClient(node.EsploraUrl(), network) // ANCHOR: ignore
	if err != nil {                                                // ANCHOR: ignore
		panic(err) // ANCHOR: ignore
	} // ANCHOR: ignore

	update, err := client.FullScan(wollet)
	if err != nil {
		log.Fatal(err)
	}
	if update != nil {
		if err := wollet.ApplyUpdate(*update); err != nil {
			log.Fatal(err)
		}
	}
	// ANCHOR_END: client

	nodeAddress := node.GetNewAddress()
	sendSats := uint64(1000)
	lbtc := policyAsset

	// ANCHOR: tx
	b := network.TxBuilder()
	if err := b.AddRecipient(nodeAddress, sendSats, lbtc); err != nil {
		log.Fatal(err)
	}
	pset, err := b.Finish(wollet)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: tx

	// ANCHOR: pset-details
	details, err := wollet.PsetDetails(pset)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: pset-details
	_ = details

	// ANCHOR: sign
	pset, err = signer.Sign(pset)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: sign

	// ANCHOR: broadcast
	tx, err := pset.Finalize()
	if err != nil {
		log.Fatal(err)
	}
	txid, err = client.Broadcast(tx)
	if err != nil {
		log.Fatal(err)
	}

	// optional
	if err := wollet.ApplyTransaction(tx); err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: broadcast
	fmt.Println("Broadcast txid:", txid)
}
